apiserver/src/auth/authorizor.rs

```diff
@@ -0,0 +1,438 @@
+//! This module provides abstractions for authenticating and authorizing requests from brupop agents to make changes to
+//! their Node's resources (including BottlerocketNode custom resources, or Draining their host Nodes of Pods.)
+use super::error::*;
+use models::node::BottlerocketNodeSelector;
+
+use async_trait::async_trait;
+use k8s_openapi::api::{
+    authentication::v1::{TokenReview, TokenReviewSpec, TokenReviewStatus},
+    core::v1::Pod,
+};
+use kube::{
+    api::{Api, PostParams},
+    runtime::reflector::{ObjectRef, Store},
+};
+use snafu::OptionExt;
+use tracing::instrument;
+
+use std::collections::HashSet;
+
+/// A token authorizor can determine if a given identity is authorized to make changes to a particular node.
+#[async_trait]
+pub trait TokenAuthorizor: Clone {
+    /// Determine if the identity represented by the provided auth token has access to the provided node.
+    async fn check_request_authorized(
+        &self,
+        node_selector: &BottlerocketNodeSelector,
+        auth_token: &str,
+    ) -> Result<(), AuthorizationError>;
+}
+
+// The k8s TokenReview authenticator adds the pod name in the `extra` field of the UserInfo
+// provided with our TokenReview.status. This is the key to retrieve it.
+//
+// Unfortunately, the presence in the `extra` field means that this is non-standard, so it's possible
+// that brupop will not work with some implementations of a TokenReview server.
```

Comment:
"non-standard" sounds scary - it seems like this is guaranteed by the current Kubernetes API implementation, at any rate.

Reply:
Yes, same with audience-awareness. I don't anticipate having problems with upstream Kubernetes, but it does seem like there are hints that a nonstandard implementation could do something different.